### PR TITLE
Update get_merged_lora_ckpt  for dist checkpoints

### PR DESCRIPTION
**Pull request by ankitageorge**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* There was a bug where the get_merged_lora_ckpt method doesn't work in distributed recipes when async checkpointing was enabled, because the existing method expects the data to be all on rank-0. This PR adds a new method get_merged_lora_dist_ckpt to be used for dist checkpoints and appropriately calls it when needed
* note that this doesn't work for nf4 tensors, I've added a comment in the code as well
* When testing on 70B model with the lora_finetune_distributed recipe, it took ~20s, compared to 63s for gathering and ~250s for saving (~310s total), saving almost 5 minutes
* since we save the adapter weights separately, change the order to save that first, so that the trainer is unblocked faster. This because dist checkpointer blocks on itself for two separate calls, so if we save adapter checkpoint after normal, then we are blocking on the normal checkpoint saving time, which is longer than adapter. 

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x ] run unit tests via `pytest tests`
- [x ] run recipe tests via `pytest tests -m integration_test`
- [x ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/modules/peft/_utils.py [hunk 1]
```diff
@@ -256,6 +257,123 @@ def get_merged_lora_ckpt(
...
+        alpha (float): The alpha value used for scaling LoRA decompositions.
+    Returns:
+        dict[str, Any]: The merged state dict.
+    """
+
+    lora_modules = _get_lora_modules(state_dict)
+    lora_moe_modules = _get_lora_moe_modules(state_dict)
+
+    # Create a simple module for matrix multiplication
+    class MatMulModule(torch.nn.Module):
```
Comment (joecummings) [inline torchtune/modules/peft/_utils.py]: Why do we need this instead of just calling matmul directly?

File: torchtune/modules/peft/_utils.py [hunk 2]
```diff
@@ -256,6 +257,123 @@ def get_merged_lora_ckpt(
...
+
+    for module in sorted(lora_modules.union(lora_moe_modules)):
+        # TODO: we don't currently support DoRA for MoE layers
+        if "experts" in module:
+            for param in ["gate", "up", "down"]:
+                lora_a_weight = state_dict[f"{module}.lora_{param}_a"]
+                lora_b_weight = state_dict[f"{module}.lora_{param}_b"]
+
+                # Create a simple module for transpose operation
+                class TransposeModule(torch.nn.Module):
```
Comment (joecummings) [inline torchtune/modules/peft/_utils.py]: Same here: why does this need to be a transpose module?

File: torchtune/modules/peft/_utils.py (hunk 1 above)
Comment (ankitageorge) [reply torchtune/modules/peft/_utils.py]: these operations don't work properly on d-tensors. it's what was causing the hangs

**Review by joecummings — APPROVED** [review]

File: torchtune/modules/peft/_utils.py (hunk 1 above)
Comment (joecummings) [reply torchtune/modules/peft/_utils.py]: "doesn't work properly" - can you expand on that?

Comment (ankitageorge) [reply torchtune/modules/peft/_utils.py]: ok actually, I think it's not needed. Good catch. I thought it was causing problems, but I just re-tested without it and it still works. I'll get rid of them and just add barriers to the existing method.
